File: Difficulty/Hard/[25] Reverse Nodes in k-Group/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode *reverseKGroup(ListNode *head, int k) {
        if (! head || ! head->next) return head;
        int num = 0;
        ListNode* res = new ListNode(0);
        res->next = head;
        ListNode *pre = res, *cur = res, *nex;
        while(cur = cur->next) num += 1;
        while(num >= k){
            cur = pre->next;
            nex = cur->next;
            for (int i = 1; i < k; ++i){
                cur->next = nex->next;
                nex->next = pre->next;
                pre->next = nex;
                nex = cur->next;
            }
            pre = cur;
            num -= k;
        }
        return res->next;
    }
};

File: Difficulty/Hard/[25] Reverse Nodes in k-Group/solution.py
#
# @lc app=leetcode id=25 lang=python3
#
# [25] Reverse Nodes in k-Group
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next:
            return head
        res = ListNode(0)
        res.next = head
        nex = ListNode(0)
        pre = cur = res
        num = 0
        while(cur):
            cur = cur.next
            num += 1
        while(num > k):
            cur = pre.next
            nex = cur.next
            for _ in range(1, k):
                cur.next = nex.next
                nex.next = pre.next
                pre.next = nex
                nex = cur.next
            pre = cur
            num -= k
        return res.next